router/manager/manager.go

```diff
@@ -0,0 +1,162 @@
+	}
+	r.brt = &batchrouter.Factory{
+		Reporting:     r.reportingI,
+		Multitenant:   multitenant.NOOP,
+		BackendConfig: r.backendConfig,
+		RouterDB:      &r.DBs.BatchRouterDB,
+		ProcErrorDB:   &r.DBs.ProcErrDB,
+	}
+
+	currentCtx, cancel := context.WithCancel(context.Background())
+	r.currentCancel = cancel
+	g, _ := errgroup.WithContext(context.Background())
+	r.waitGroup = g
+	g.Go(func() error {
+		r.monitorDestRouters(currentCtx, *r.rt, *r.brt)
+		return nil
+	})
+}
+
+// Stop stops the Router, this is a blocking call.
+func (r *Router) Stop() {
+	r.currentCancel()
+	r.waitGroup.Wait()
+}
+
+// NewRouterManager creates a new Router instance
+func NewRouterManager(ctx context.Context, dbs *jobsdb.DBs) *Router {
+	router.RoutersManagerSetup()
+	batchrouter.BatchRoutersManagerSetup()
+
+	return &Router{
+		rt:               &router.Factory{},
+		brt:              &batchrouter.Factory{},
+		mainCtx:          ctx,
+		DBs:              dbs,
+		multitenantStats: multitenant.NOOP,
+		backendConfig:    backendconfig.DefaultBackendConfig,
+	}
+}
+
+// Gets the config from config backend and extracts enabled writekeys
+func (r *Router) monitorDestRouters(ctx context.Context, routerFactory router.Factory,
+	batchrouterFactory batchrouter.Factory) {
+	ch := make(chan utils.DataEvent)
+	r.backendConfig.Subscribe(ch, backendconfig.TopicBackendConfig)
+	dstToRouter := make(map[string]*router.HandleT)
+	dstToBatchRouter := make(map[string]*batchrouter.HandleT)
+	cleanup := make([]func(), 0)
+
+	//Crash recover routerDB, batchRouterDB
+	//Note: The following cleanups can take time if there are too many
+	//rt / batch_rt tables and there would be a delay readin from channel `ch`
+	//However, this shouldn't be the problem since backend config pushes config
+	//to its subscribers in separate goroutines to prevent blocking.
+	routerFactory.RouterDB.DeleteExecuting(jobsdb.GetQueryParamsT{JobCount: -1})
+	batchrouterFactory.RouterDB.DeleteExecuting(jobsdb.GetQueryParamsT{JobCount: -1})
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break loop
+		case config := <-ch:
+			sources := config.Data.(backendconfig.ConfigT)
+			enabledDestinations := make(map[string]bool)
+			for _, source := range sources.Sources {
+				for _, destination := range source.Destinations {
+					enabledDestinations[destination.DestinationDefinition.Name] = true
+					//For batch router destinations
+					if misc.ContainsString(objectStorageDestinations, destination.DestinationDefinition.Name) || misc.ContainsString(warehouseDestinations, destination.DestinationDefinition.Name) || misc.ContainsString(asyncDestinations, destination.DestinationDefinition.Name) {
+						_, ok := dstToBatchRouter[destination.DestinationDefinition.Name]
+						if !ok {
+							pkgLogger.Info("Starting a new Batch Destination Router ", destination.DestinationDefinition.Name)
+							brt := batchrouterFactory.New(destination.DestinationDefinition.Name)
+							brt.Start()
```

Comment:
Not in the current scope, But can we Start these brt and rt in parallel? This would immensly reduce the start up time on hosted

Reply:
Noted